{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Studies have shown that gatekeeper  mutants  affect the drug target binding of FGF receptor tyrosine kinases.  Your task are the following:\n",
    "\n",
    "- Create a function (A) that accepts a file as an argument (.5pts).\n",
    "- Within the function open the file for reading (.5pts).\n",
    "- Within the function create a new file for writing output (.5pts).\n",
    "- Return the fasta sequences with 'fibroblast growth factor receptor 1 isoform' in its definition line (.5pts).\n",
    "- Create another function (B) that will read the output from Function A (.5pts).  \n",
    "- Within Function B, determine if position 564 has a 'V', 'I,' or 'E'. Valine is the wild type and the gatekeeper mutation at this position is either Isoleucine (I) or Glutamic Acid (E) (.5pts).\n",
    "- The final output should be the attached example file. One is for yes and zero means no (.5pts).\n",
    "- Chaining the analysis as pipeline 1.5pts."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**My Thoughts**:\n",
    "\n",
    "Function A:\n",
    "- needs to accept a file as an argument\n",
    "- within the file open a the file for reading open('r')\n",
    "- open the file for writing open('w')\n",
    "- if 'fibroblast growth factor receptor 1' is in line, add to outfile\n",
    "\n",
    "Function B:\n",
    "- will take the outfile produced from A\n",
    "- organize into a dictionary\n",
    "- if position 564 is V, I or E, 1 for yes, 0 for no."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "inPath = r\"C:\\Users\\tiffa\\OneDrive\\Desktop\\Masters in Bioinformatics\\Problem Solving\\Homeworks\\FGFR_fusion.fasta\"\n",
    "#path for the fusion fasta file copied directly from my directory\n",
    "#I need to make my path a raw literal string or else it throws errors\n",
    "outPath = r\"C:\\Users\\tiffa\\OneDrive\\Desktop\\Masters in Bioinformatics\\Problem Solving\\Homeworks\\FilesHWout.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def function_A(input_file):\n",
    "    outFile = outPath #create file for writing\n",
    "    out = open(outFile, 'w')\n",
    "    #define my out file object as a raw literal string, or else it throws errors\n",
    "    with open(input_file, 'r') as f_input:\n",
    "        for line in f_input:\n",
    "            #for each line in our read in file.\n",
    "            if 'fibroblast growth factor receptor 1 isoform' in line:\n",
    "                #if the line has the key words, we will record it and it's sequence\n",
    "                while not line.startswith('\\n'):\n",
    "                    #I am using starting with a newline as a delimeter\n",
    "                    #so it should write all lines up to an empty line\n",
    "                    out.write(line)\n",
    "                    line = next(f_input)\n",
    "                    #set the line to be the next line\n",
    "    out.close()\n",
    "    #remember to close out the file when you're done writing!\n",
    "    function_B(outFile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#I am going to keep a dictionary to store my fasta sequences. \n",
    "def function_B(output_file):\n",
    "\n",
    "    fastaSeq = {} #initialize my empty dictionary\n",
    "    #this sequence will populate my dictionary for easier data analysis\n",
    "\n",
    "    with open(output_file, 'r') as f:\n",
    "        #open and read my file\n",
    "        header = ''#initialize empty header variable\n",
    "        sequence = ''#initialize empty sequence variable\n",
    "        \n",
    "        for line in f:#iterate over each line in the file\n",
    "            \n",
    "            #line = line.strip()#clean up the line of white spaces and newline chars\n",
    "            #it's position 563 with the newline characters\n",
    "            if line.startswith('>'):\n",
    "                if header:  # If header is not empty (i.e., not the first header)\n",
    "                    fastaSeq[header] = sequence  # Store the previous header and sequence\n",
    "                header = line[0:]  # Extract header without '>'\n",
    "                sequence = ''  # Reset sequence for the new header\n",
    "            else:\n",
    "                sequence += line  # Concatenate lines to form the sequence\n",
    "        # Store the last header and sequence\n",
    "        if header:\n",
    "            fastaSeq[header] = sequence\n",
    "\n",
    "    #now the dictionary should be set, we can go through our dictionary, open the file as write\n",
    "    out = open(output_file, 'w')#open the file as writing mode\n",
    "    out.write('ID\\tValine\\tIsoleucine\\tGlutamic Acid\\n')\n",
    "    #put a header into our out file\n",
    "    for header, sequence in fastaSeq.items(): #for each pair in our dictionary\n",
    "            header = header.strip() #clean up the header of spaces and blank characters before writing\n",
    "            out.write(header) #write the header\n",
    "            AA = sequence[563] #check position 564, python is 0 index based, so it would be index 563\n",
    "            if AA == 'V':\n",
    "                 #if it is a V, put a 1 in the first position\n",
    "                 out.write(' 1\\t0\\t0\\n')\n",
    "\n",
    "            elif AA == 'I':\n",
    "                 #if it is an I, put a 1 in the second position\n",
    "                 out.write(' 0\\t1\\t0\\n')\n",
    "\n",
    "            elif AA == 'E':\n",
    "                #if it is an E, put a 1 in the third position\n",
    "                out.write(' 0\\t0\\t1\\n')\n",
    "\n",
    "            else:\n",
    "                #put 0 for all positions\n",
    "                out.write(' 0\\t0\\t0\\n')\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "function_A(inPath)\n",
    "#functions are chained together, so calling function A will result in function B's output\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MWSWKCLLFWAVLVTATLCTARPSPTLPEQAQPWGAPVEVESFLVHPGDLLQLRCRLRDDVQSINWLRDGVQLAESNRTRITGEEVEVQDSVPADSGLYACVTSSPSGSDTTYFSVNVSDALPSSEDDDDDDDSSSEEKETDNTKPNPVAPYWTSPEKMEKKLHAVPAAKTVKFKCPSSGTPNPTLRWLKNGKEFKPDHRIGGYKVRYATWSIIMDSVVPSDKGNYTCIVENEYGSINHTYQLDVVERSPHRPILQAGLPANKTVALGSNVEFMCKVYSDPQPHIQWLKHIEVNGSKIGPDNLPYVQILKTAGVNTTDKEMEVLHLRNVSFEDAGEYTCLAGNSIGLSHHSAWLTVLEALEERPAVMTSPLYLEIIIYCTGAFLISCMVGSVIVYKMKSGTKKSDFHSQMAVHKLAKSIPLRRQVTVSADSSASMNSGVLLVRPSRLSSSGTPMLAGVSEYELPEDPRWELPRDRLVLGKPLGEGCFGQVVLAEAIGLDKDKPNRVTKVAVKMLKSDATEKDLSDLISEMEMMKMIGKHKNIINLLGACTQDGPLYVIVEYASKGNLREYLQARRPPGLEYCYNPSHNPEEQLSSKDLVSCAYQVARGMEYLASKKCIHRDLAARNVLVTEDNVMKIADFGLARDIHHIDYYKKTTNGRLPVKWMAPEALFDRIYTHQSDVWSFGVLLWEIFTLGGSPYPGVPVEELFKLLKEGHRMDKPSNCTNELYMMMRDCWHAVPSQRPTFKQLVEDLDRIVALTSNQEYLDLSMPLDQYSPSFPDTRSSTCSSGEDSVFSHEPLPEEPCLPRHPAQLANGGLKRR\n",
      "K\n",
      "V\n",
      "match V, 11\n",
      "match V, 13\n",
      "match V, 37\n",
      "match V, 39\n",
      "match V, 44\n",
      "match V, 60\n",
      "match V, 71\n",
      "match V, 86\n",
      "match V, 88\n",
      "match V, 92\n",
      "match V, 102\n",
      "match V, 116\n",
      "match V, 118\n",
      "match V, 150\n",
      "match V, 167\n",
      "match V, 173\n",
      "match V, 207\n",
      "match V, 220\n",
      "match V, 221\n",
      "match V, 232\n",
      "match V, 247\n",
      "match V, 248\n",
      "match V, 267\n",
      "match V, 273\n",
      "match V, 279\n",
      "match V, 296\n",
      "match V, 309\n",
      "match V, 317\n",
      "match V, 326\n",
      "match V, 332\n",
      "match V, 360\n",
      "match V, 370\n",
      "match V, 393\n",
      "match V, 396\n",
      "match V, 398\n",
      "match V, 416\n",
      "match V, 430\n",
      "match V, 432\n",
      "match V, 444\n",
      "match V, 447\n",
      "match V, 463\n",
      "match V, 482\n",
      "match V, 495\n",
      "match V, 497\n",
      "match V, 512\n",
      "match V, 515\n",
      "match V, 517\n",
      "match V, 563\n",
      "match V, 565\n",
      "match V, 606\n",
      "match V, 612\n",
      "match V, 634\n",
      "match V, 636\n",
      "match V, 642\n",
      "match V, 670\n",
      "match V, 689\n",
      "match V, 694\n",
      "match V, 711\n",
      "match V, 713\n",
      "match V, 747\n",
      "match V, 758\n",
      "match V, 765\n",
      "match V, 803\n"
     ]
    }
   ],
   "source": [
    "#my original output was not yielding any results, so I looked through the positions\n",
    "#in the example you gave, in the output example it had a V in position 564, \n",
    "#so here I am checking if position 564 with newlines, and without newlines\n",
    "import re\n",
    "seq = \"\"\"MWSWKCLLFWAVLVTATLCTARPSPTLPEQAQPWGAPVEVESFLVHPGDLLQLRCRLRDDVQSINWLRDG\n",
    "VQLAESNRTRITGEEVEVQDSVPADSGLYACVTSSPSGSDTTYFSVNVSDALPSSEDDDDDDDSSSEEKE\n",
    "TDNTKPNPVAPYWTSPEKMEKKLHAVPAAKTVKFKCPSSGTPNPTLRWLKNGKEFKPDHRIGGYKVRYAT\n",
    "WSIIMDSVVPSDKGNYTCIVENEYGSINHTYQLDVVERSPHRPILQAGLPANKTVALGSNVEFMCKVYSD\n",
    "PQPHIQWLKHIEVNGSKIGPDNLPYVQILKTAGVNTTDKEMEVLHLRNVSFEDAGEYTCLAGNSIGLSHH\n",
    "SAWLTVLEALEERPAVMTSPLYLEIIIYCTGAFLISCMVGSVIVYKMKSGTKKSDFHSQMAVHKLAKSIP\n",
    "LRRQVTVSADSSASMNSGVLLVRPSRLSSSGTPMLAGVSEYELPEDPRWELPRDRLVLGKPLGEGCFGQV\n",
    "VLAEAIGLDKDKPNRVTKVAVKMLKSDATEKDLSDLISEMEMMKMIGKHKNIINLLGACTQDGPLYVIVE\n",
    "YASKGNLREYLQARRPPGLEYCYNPSHNPEEQLSSKDLVSCAYQVARGMEYLASKKCIHRDLAARNVLVT\n",
    "EDNVMKIADFGLARDIHHIDYYKKTTNGRLPVKWMAPEALFDRIYTHQSDVWSFGVLLWEIFTLGGSPYP\n",
    "GVPVEELFKLLKEGHRMDKPSNCTNELYMMMRDCWHAVPSQRPTFKQLVEDLDRIVALTSNQEYLDLSMP\n",
    "LDQYSPSFPDTRSSTCSSGEDSVFSHEPLPEEPCLPRHPAQLANGGLKRR\"\"\"\n",
    "joined = seq.replace('\\n', '')\n",
    "print(joined)\n",
    "looking = r'V'\n",
    "print(joined[563])\n",
    "print(seq[563])\n",
    "#it's position 564 with the newline characters, \n",
    "matches = re.finditer(looking, seq)\n",
    "#it's position 564 with the newline characters, \n",
    "for match in matches:\n",
    "    print(f'match {match.group()}, {match.start()}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above testing showed that the sample output file given was produced after analyzing the results WITH newline characters left in. I amended my code so that it would look for the 564th position WITH newlines because of this, as I had originally looked for what is in the 564th position WITHOUT newlines. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
